### PR TITLE
Load clinical and research gene fusion reports on case upload or load/update them later using the CLI

### LINKED ISSUE #2343: Add transcript fusion report link to reports section in existing case page
**Background**
When analysing an individual it is common to start with a DNA analysis. However, often a causative variant is not found and the same individual also are subjected to a RNA analysis. 

**Challenge**
Currently, there is no way to combine the output from the standard DNA analysis with the selected output from a RNA analysis in Scout. It would be great to be able to update the DNA case view with information on transcript fusions in the form of a pdf report that can be viewed/downloaded from Scout.

**Suggested solution**
- Add the option in `scout case update` to also be able to upload a link pointing to a transcript fusion report from a secondary analysis to an already existing case.
- Also support this from the scout load config
- Add the optional transcript fusion report to the reports section in an already existing DNA analysis case.

### PR DESCRIPTION
**Pull request by northwestwitch**
Fix #2343.

- This PR allows loading gene fusion reports (clinical and research) along with other info on case load, when the key/value 'gene_fusion_report' and 'gene_fusion_report_research' is present in the Scout yaml config file.
- Allows also to add/update the report of an existing case in the database by running the command: `scout load gene-fusion-report`:

image

  

**How to test**:
- [ ] Test on the demo case by running `scout setup demo` or uploading the case again with `scout --demo load case path_to_config_file -u`
- [ ] Try to update the clinical fusion report with any PDF of your choice by following the cli instructions
- [ ] Try to load a research fusion report (any PDF of your choice) from the command line by following the cli instructions
- [ ] Make sure that the links to these 2 reports are correctly visualized on the case sidebar and that clicking on them shows the files

**Expected outcome**:
1. The demo case should show a link to the fusion report on the side bar. Clicking on this link should download the report, given that the file exists in that path on the disk

**Review:**
- [ ] code approved by
- [ ] tests executed by

### CODE REVIEW

**Comment by northwestwitch** [comment]
This is ready for review

**Comment by northwestwitch** [comment]
image

And when you click on the link you get this:

image

**Comment by henrikstranneheim** [comment]
Beautiful!

**Comment by northwestwitch** [comment]
> Beautiful!

Thanks I've copied it from Arriba. But is it like this for real?

**Comment by henrikstranneheim** [comment]
> > Beautiful!
> 
> Thanks I've copied it from Arriba. But is it like this for real?

yes

**Comment by northwestwitch** [comment]
Ok, fixed this PR with demo files I got from @jemten. Ready for review again!

**Comment by jemten** [comment]
Oh, I just struck me. There will be one report per sample in the case. Sorry if this messes things up

**Comment by northwestwitch** [comment]
> Oh, I just struck me. There will be one report per sample in the case. Sorry if this messes things up

OK, can't we somehow merge these reports in a unique pdf?

**Comment by jemten** [comment]
Possibly. But since the reports doesn't contain any sample names to indicate which fusion belongs to which sample it won't be as easy as just concatenating them.

**Comment by northwestwitch** [comment]
> Possibly. But since the reports doesn't contain any sample names to indicate which fusion belongs to which sample it won't be as easy as just concatenating them.

I see. Then it's more complicated, yes. Let's think about it!

**Comment by northwestwitch** [comment]
I'm working on a program that will merge and bookmark the files after writing on them the name of the original PDF. Hopefully I'll be finished with it later today

**Comment by henrikstranneheim** [comment]
> I'm working on a program that will merge and bookmark the files after writing on them the name of the original PDF. Hopefully I'll be finished with it later today

Cool! Should we run this in MIP in the arriba step or where are you planning to inject it?

**Comment by northwestwitch** [comment]
> Cool! Should we run this in MIP in the arriba step or where are you planning to inject it?

I think it should be part of the pipeline, before the injection. How do you prefer to feed the pdf files to the program:
- invoke the program and give it the list of all files to merge
- invoke the program with the directory where these files are located (in this case it will merge everything that finds with a .pdf suffix, after ordering them by name)

?

**Comment by henrikstranneheim** [comment]
Let's go with supplying a list of files to merge via the CLI

**Comment by jemten** [comment]
> Let's go with supplying a list of files to merge via the CLI

I agree

**Comment by northwestwitch** [comment]
@henrikstranneheim @jemten  this one: https://github.com/northwestwitch/pdfmerger

I could use some tester with the arriba files and some improvement suggestion!

**Comment by jemten** [comment]
Great work @northwestwitch. 
I'll check it out and add it to the pipeline

**Comment by northwestwitch** [comment]
@jemten @henrikstranneheim I've created a Dockerfile to be pulled and runned on the fly to merge outfiles from arriba (or any other pdf report).

Command:

```
docker run -d --rm \
-v /home/username/documents:/home/pdfmerger/data \
northwestwitch/pdfmerger --orientation landscape \
-f /home/pdfmerger/data/infile1.pdf -f /home/pdfmerger/data/infile2.pdf \
--outfolder /home/pdfmerger/data (--outfile outfilename.pdf)
```

See more info [here](https://github.com/northwestwitch/pdfmerger). 

By default the program doesn't write the file names on each page of each file before merging. If you wish to add the names you can use the additional tag `--write-filenames`

**Comment by northwestwitch** [comment]
This is now fixed according to the discussion we had yesterday @henrikstranneheim and @jemten : 
- You can load clinical gene fusion report and research gene fusion report directly when you load the case (keys: `gene_fusion_report` and `gene_fusion_report_research`)
- You can update a case with the path one of the 2 reports successively. 

On the case page:

image

File: scout/build/case.py
Comment (dnil) [inline scout/build/case.py:4]: 👍

File: scout/commands/load/base.py
Comment (dnil) [inline scout/commands/load/base.py:41]: We probably want a load_custom_report at some point (major =), but this is perfect now!

File: scout/server/blueprints/cases/templates/cases/collapsible_actionbar.html
Comment (dnil) [inline scout/server/blueprints/cases/templates/cases/collapsible_actionbar.html:500]: 😸

File: scout/server/blueprints/cases/views.py
Comment (dnil) [inline scout/server/blueprints/cases/views.py:880]: This is not really that dangerous, I think - as it is sort of limited to what is on the case_obj. And we have not really gone over everything like it, but since Im reading. I suppose you could if you really, really wanted add a path into like the synopsis or tissue type fields, then retrieve it by asking for report_type="synopsis" or such. Perhaps limit it to the valid keys, eg the report ones from constants?

**Review by dnil — APPROVED** [review]

File: scout/commands/load/base.py
Comment (northwestwitch) [reply scout/commands/load/base.py:41]: I totally agree, it would be so nice to generalize and allow people to provide and visualize any report they want. We could already start thinking about it

File: scout/server/blueprints/cases/views.py
Comment (northwestwitch) [reply scout/server/blueprints/cases/views.py:880]: Especially true if we start generalizing the available reports. Very good suggestion, I'll fix it!

**Comment by dnil** [comment]
Did you want a test on this as well? Kind of looks like that is already done above, but if you make a minor change now, I could test afterwards.

**Comment by northwestwitch** [comment]
> Did you want a test on this as well? Kind of looks like that is already done above, but if you make a minor change now, I could test afterwards.

I'll do that change in the gene_fusion_report view. Feel free to give it a try later, thanks!

**Comment by northwestwitch** [comment]
Tested after the change, it works: 

image

File: scout/server/blueprints/cases/views.py
Comment (dnil) [inline scout/server/blueprints/cases/views.py:881]: 💯

**Review by dnil — APPROVED** [review]

**Pull request merged**
